src/main/flight/position.c

```diff
@@ -22,205 +22,202 @@
-            if (haveBaroAlt && positionConfig()->altitude_source == DEFAULT) {
-                // mix zeroed GPS with Baro altitude data, if Baro data exists if are in default altitude control mode 
-                const float absDifferenceM = fabsf(zeroedAltitudeCm - baroAltCm) / 100.0f * positionConfig()->altitude_prefer_baro / 100.0f;
-                if (absDifferenceM > 1.0f) { // when there is a large difference, favour Baro
-                    gpsTrust /=  absDifferenceM;
+        switch (positionConfig()->altitude_source) {
+            case ALTITUDE_SOURCE_DEFAULT:
+#if defined(USE_GPS) && defined(USE_BARO)
+                // Fuse GPS and baro for more precise altitude
+                if (isGpsAvailable && isBaroAvailable) {  // Complementary filter
+                    gpsAltLowpassCm = pt1FilterApply(&gpsFuseLpf, gpsAltCm - gpsAltGroundCm);
+                    baroAltHighpassCm = baroAltCm - pt1FilterApply(&baroFuseLpf, baroAltCm);
+                    altitudeCm = gpsAltLowpassCm + baroAltHighpassCm;
+                } else
+#endif
+                if (isGpsAvailable && !isBaroAvailable) {  // Fallback option if baro fails
+                    altitudeCm = gpsAltCm - gpsAltGroundCm;
+                } else if (!isGpsAvailable && isBaroAvailable) {  // Fallback option if GPS fails
+                    altitudeCm = baroAltCm;
                 }
-                zeroedAltitudeCm = zeroedAltitudeCm * gpsTrust + baroAltCm * (1.0f - gpsTrust);
-            }
-        } else if (haveBaroAlt && (positionConfig()->altitude_source == DEFAULT || positionConfig()->altitude_source == BARO_ONLY)) {
-            zeroedAltitudeCm = baroAltCm; // use Baro if no GPS data, or we want Baro only
+                break;
+            case ALTITUDE_SOURCE_GPS_ONLY:
+                altitudeCm = gpsAltCm - gpsAltGroundCm;
+                break;
+            case ALTITUDE_SOURCE_BARO_ONLY:
+                altitudeCm = baroAltCm;
+                break;
+            default:
+                break;
         }
     }
 
-    zeroedAltitudeCm = pt2FilterApply(&altitudeLpf, zeroedAltitudeCm);
-    // NOTE: this filter must receive 0 as its input, for the whole disarmed time, to ensure correct zeroed values on arming
-
     if (wasArmed) {
-        displayAltitudeCm = zeroedAltitudeCm; // while armed, show filtered relative altitude in OSD / sensors tab
+        displayAltitudeCm = altitudeCm; // while armed, show filtered relative altitude in OSD / sensors tab
     }
 
-    // *** calculate Vario signal
-    static float previousZeroedAltitudeCm = 0.0f;
-    zeroedAltitudeDerivative = (zeroedAltitudeCm - previousZeroedAltitudeCm) * TASK_ALTITUDE_RATE_HZ; // cm/s
-    previousZeroedAltitudeCm = zeroedAltitudeCm;
+#ifdef USE_VARIO
+    // calculate vario signal
+    static float prevAltitudeCm = 0.0f;
+    climbRateCmS = (altitudeCm - prevAltitudeCm) * TASK_ALTITUDE_RATE_HZ; // cm/s
```

Comment:
Use temporary here (don't write unfiltered values into climbRateCmS)

Reply:
For compiler optimization?

Reply:
compiler optimization makes this change zero cost - generated code will be probably the same. 
The idea is to make climbRateCmS always valid, even in ISR or different thread.